mate-panel/wayland-backend.c

```diff
@@ -0,0 +1,755 @@
+				popup_data,
+				(GDestroyNotify) wayland_popup_data_destroy);
+
+		// On unmap, we need to destroy the shell surface we create before GTK destroys its wl_surface
+		// To do that, we have to override the default unmap signal
+		wayland_popup_override_unmap_signal (popup_widget);
+		g_signal_connect (popup_widget, "realize", G_CALLBACK (wayland_popup_realize_cb), NULL);
+		g_signal_connect (popup_widget, "map-event", map_event_cb, popup_data);
+	}
+
+	popup_data->widget = popup_widget;
+	popup_data->attach_widget = attach_widget;
+}
+
+static gboolean
+wayland_menu_map_event_cb (GtkWidget *popup_widget, GdkEvent *event, WaylandPopupData *popup_data)
+{
+	WaylandPopupData *parent_popup_data;
+	PanelToplevel *toplevel;
+	enum xdg_positioner_anchor anchor = XDG_POSITIONER_ANCHOR_TOP_LEFT;
+	enum xdg_positioner_gravity gravity = XDG_POSITIONER_GRAVITY_BOTTOM_RIGHT;
+	GdkPoint offset = {0, 0};
+
+	g_return_val_if_fail (popup_data, FALSE);
+	g_return_val_if_fail (popup_data->widget == popup_widget, FALSE);
+
+	parent_popup_data = g_object_get_data (G_OBJECT (gtk_widget_get_toplevel (popup_data->attach_widget)),
+					       wayland_popup_data_key);
+	toplevel = wayland_popup_data_get_panel_toplevel (popup_data);
+
+	if (parent_popup_data) {
+		// We assume popups on popups behave like nested menus
+		anchor = XDG_POSITIONER_ANCHOR_TOP_RIGHT;
+		gravity = XDG_POSITIONER_GRAVITY_BOTTOM_RIGHT;
+	} else if (toplevel) {
+		switch (panel_toplevel_get_orientation (toplevel)) {
+		case PANEL_ORIENTATION_TOP:
+			anchor = XDG_POSITIONER_ANCHOR_BOTTOM_LEFT;
+			gravity = XDG_POSITIONER_GRAVITY_BOTTOM_RIGHT;
+			break;
+		case PANEL_ORIENTATION_RIGHT:
+			anchor = XDG_POSITIONER_ANCHOR_TOP_LEFT;
+			gravity = XDG_POSITIONER_GRAVITY_BOTTOM_LEFT;
+			break;
+		case PANEL_ORIENTATION_BOTTOM:
+			anchor = XDG_POSITIONER_ANCHOR_TOP_LEFT;
+			gravity = XDG_POSITIONER_GRAVITY_TOP_RIGHT;
+			break;
+		case PANEL_ORIENTATION_LEFT:
+			anchor = XDG_POSITIONER_ANCHOR_TOP_RIGHT;
+			gravity = XDG_POSITIONER_GRAVITY_BOTTOM_RIGHT;
+			break;
+		}
+	} else {
+		g_warning ("Orientation unknown because menu has not toplevel");
+	}
+
+	if (popup_data->attach_widget == GTK_WIDGET (toplevel)) {
+		anchor = XDG_POSITIONER_ANCHOR_TOP_LEFT;
+		widget_get_pointer_position (popup_data->attach_widget, &offset.x, &offset.y);
+	}
+
+	wayland_pop_popup_up_at_widget (popup_data,
+					anchor,
+					gravity,
+					offset);
+
+	return TRUE;
+}
+
+void
+wayland_popup_menu_setup (GtkWidget *menu, GtkWidget *attach_widget)
+{
+	wayland_setup_popup_data (gtk_widget_get_toplevel (menu),
+				  attach_widget,
+				  G_CALLBACK (wayland_menu_map_event_cb));
+}
+
+static gboolean
+wayland_tooltip_map_event_cb (GtkWidget *popup_widget, GdkEvent *event, WaylandPopupData *popup_data)
+{
+	PanelToplevel *toplevel;
+	enum xdg_positioner_anchor anchor = XDG_POSITIONER_ANCHOR_TOP_LEFT;
+	enum xdg_positioner_gravity gravity = XDG_POSITIONER_GRAVITY_BOTTOM_RIGHT;
+	static const int gap = 6;
```

Comment:
What is this magic number?

Reply:
It controls the placement of tooltips. I see it's unclear. I'll move it to be a better named and commented constant at the top.